src/coreclr/jit/emitxarch.cpp

```diff
@@ -13080,7 +13080,7 @@ BYTE* emitter::emitOutputRI(BYTE* dst, instrDesc* id)
                 break;
 
             case IF_RRW_CNS:
-                assert(id->idGCref() == GCT_BYREF);
+                assert((id->idGCref() == GCT_BYREF) || (id->idGCref() == GCT_GCREF));
```

Comment:
Looks like the `emitGCregLiveUpd(GCT_BYREF, id->idReg1(), dst);` call below also needs to be fixed.

But it seems strange to me. How are we producing a gcref after a `r/w reg, const` operation? Seems like we are either misreporting this or the const has to be 0 and the instruction is useless.

Reply:
I've not investigated this one yet (where it come from, I assume it's jitstressregs mode) but it looks like this:
```
IN0006: 000030 B901000000           mov      ecx, 1
IN0007: 000035 48BA0800FDD748020000 mov      rdx, 0x248D7FD0008
                            ; gcrRegs +[rdx]
IN0008: 00003F 4881C2F8090000       add      rdx, 0x9F8
                            ; gcrRegs -[rdx]
                            ; byrRegs +[rdx]
```
so from my understanding it should not be fixed to GCREF, should it?

Reply:
> so from my understanding it should not be fixed to GCREF, should it?

I agree, but I think we are creating an incorrect instruction then. `id->idGCref() == GCT_BYREF` should be true here. Consider
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static void Foo(C c)
{
    Bar(ref c.F);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Bar(ref int i)
{
}

private class C
{
    public int F;
}
```
that produces
```asm
add rcx, 8
```
without hitting the assert today.

Reply:
@jakobbotsch thanks, I've found the reason - it turns out the test sets `DOTNET_JitConstCSE=3` so then it picks up gc const handle as a "shared constant" (it's where it may use it as a base for other constants) - that's how that `add` appeared
```

We have shared Const CSE's and selected $1c1 with a value of 0x0000024CB7AC0008 as the base.

CSE #01 def at [000897] replaced in BB02 with def of V61
ReMorphing args for 533.CALL:
Args for [000533].CALL after fgMorphArgs:
CallArg[[001079].COMMA ref (By value), 1 reg: rcx, byteAlignment=8, isLate, processed]
CallArg[[000898].CNS_INT ref (By value), 1 reg: rdx, byteAlignment=8, isLate, processed]
OutgoingArgsStackSize is 32

optValnumCSE morphed tree:
N009 ( 22, 33)              [000533] -ACXG------                         *  CALL      void   System.Diagnostics.Debug.Fail $VN.Void
N007 (  5, 16)              [001079] -A--------- arg0 in rcx             +--*  COMMA     ref    $1c0
N003 (  2, 10) CSE #01 (def)[001075] -A------R--                         |  +--*  ASG       ref    $VN.Void
N002 (  1,  1)              [001074] D------N---                         |  |  +--*  LCL_VAR   ref    V61 cse1          $VN.Void
N001 (  2, 10)              [001073] -----------                         |  |  \--*  CNS_INT   ref    0x24cb7ac0008 $1c1
N006 (  3,  6)              [001078] -------N---                         |  \--*  ADD       ref    $1c0
N004 (  1,  1)              [001076] -----------                         |     +--*  LCL_VAR   ref    V61 cse1          $1c1
N005 (  1,  4)              [001077] -----------                         |     \--*  CNS_INT   ref    0x9a8
N008 (  3, 10) CSE #01 (use)[000898] H---------- arg1 in rdx             \--*  CNS_INT(h) ref    0xB7AC0008 [ICON_STR_HDL] $1c1
```